torch/distributed/fsdp/_state_dict_utils.py

```diff
@@ -26,48 +27,217 @@
     module._assert_state([FSDP.TrainingState_.SUMMON_FULL_PARAMS])
     # Return early for trivial cases
     if not state_dict or not module._has_params:
+        _exit_full_param_ctx(module)
         return state_dict
 
-    # If a rank has already exited the `summon_full_params()` context here
-    # (e.g. when `rank0_only=True` and `rank != 0`), then the rank only
-    # needed to participate in the all-gather and does not need to save the
-    # state dict. For `use_orig_params=False`, we can check this via
-    # `FlatParameter` registration.
-    # TODO: For `use_orig_params=True`, we check for the reshard upon
-    # exiting `summon_full_params()` via the parameter shape. However, for
-    # `NO_SHARD`, we cannot tell from the shape, so we do not return early.
-    if (
-        not module._use_orig_params and FSDP.FLAT_PARAM in module.module._parameters
-    ) or (
-        module._use_orig_params
-        and module._handles
-        and module._handles[0].uses_sharded_strategy
-        and module._handles[0].is_sharded(module._handles[0].flat_param)
-    ):
-        return state_dict
+    # TODO: Once pre_state_dict hook is supported, this pop should be removed.
+    # For `use_orig_params=True`, the `FlatParameter` is not registered, so
+    # there is no entry in the state dict for it to pop.
+    if not module._use_orig_params:
+        state_dict.pop(f"{prefix}{FSDP.FLAT_PARAM}")
 
-    offload_to_cpu = module._state_dict_config.offload_to_cpu
-    cpu_device = torch.device("cpu")
+    # If a rank does not have unsharded parameters(when `rank0_only=True`
+    # and `rank != 0`), then the rank only needed to participate in the
+    # all-gather and does not need to save the # state dict. We simply check
+    # rank0_only to ensure this issue.
+    rank0_only = (
+        module._state_dict_type == FSDP.StateDictType.FULL_STATE_DICT
+        and cast(FSDP.FullStateDictConfig, module._state_dict_config).rank0_only
+    )
+    # no_fsdp_return means the state_dict returned by this rank should contain
+    # only non-FSDP controlled parameters and buffers.
+    no_fsdp_return = rank0_only and module.rank != 0
+    if no_fsdp_return and not module._use_orig_params:
+        for clean_key in module._buffer_names:
+            # This is a hack to support activation checkpoint.
+            clean_key = clean_key.replace(
+                f"{checkpoint_wrapper._CHECKPOINT_PREFIX}.", ""
+            )
+            state_dict.pop(f"{prefix}{clean_key}", None)
+        _exit_full_param_ctx(module)
+        return state_dict
 
     # Loop only the parameters saved in this instance's wrapped module to
     # avoid processing buffers.
     for fqn, param_name, module_name in module._param_fqns:
+        # TODO: remove the parameter retrieval. See ``_full_pre_state_dict_hook``.
+        param = functools.reduce(getattr, fqn.split("."), module.module)
         fqn = f"{prefix}{fqn}"
+        if no_fsdp_return:
+            state_dict.pop(fqn)
```

Comment:
is this to remove FSDP managed parameters from the state_dict if we are rank0_only and not rank 0?

Reply:
And would we still need this once we have the proper support for pre_state_dict hook in nn.Module?

Reply:
1. Yes
2. IMO, we still need this line. I think the original rank_only implementation may not be correct. The original implementation simply set state_dict to {} if `rank_only and rank !=0`. However, this may be a problem if ignore parameters are set. In such a case, we should not simply throw away what are inside `state_dict` on other ranks.

Reply:
> IMO, we still need this line. I think the original rank_only implementation may not be correct. The original implementation simply set state_dict to {} if rank_only and rank !=0. However, this may be a problem if ignore parameters are set. In such a case, we should not simply throw away what are inside state_dict on other ranks.

Good call!